### PR TITLE
Improve error messages

### PR DESCRIPTION
**Pull request by drwpow**
## Changes

Paired with @FredKSchott to improve the error stacktraces. Improves the error display in several contexts. We still have work to do, but these were some quick wins.

**Note: this PR has a lot of lines, but it’s all cohesive:**

- In order to fix stacktraces, we have to provide sourcemaps more consistently throughout. These are a lot of changes in the plugins
- In order for Vite to resolve the stacktraces, we have to pass in URLs Vite understands. Hence a lot of the changes to `viteifyURL`
- Several tests were added, which make up a lot of the bulk here
- A few improvements to error handling was added unblocked by the other stacktrace fixes

Resolves https://github.com/snowpackjs/astro/issues/1890

Comments added for explanation of changes.

**Before**

Screen Shot 2021-11-17 at 18 02 24

**After**

Screen Shot 2021-11-17 at 15 06 04

## Testing

This needs to be tested! Unfortunately, this can’t be tested without dev server tests (build tests require all pages to be working, which makes testing errors impossible, but dev server tests can handle errors because errors happen on request)

## Docs

No docs added

### CODE REVIEW

File: packages/astro/src/core/util.ts [hunk 1]
```diff
@@ -72,6 +73,13 @@ export function resolveDependency(dep: string, astroConfig: AstroConfig) {
...
-  return `/@fs/${pathname.replace(/^\//, '')}`;
+/**
+ * Vite-ify URL
+ * Given a file URL, return an ID that matches Vite’s module graph. Needed for resolution and stack trace fixing.
+ * Must match the following format:
+ * Linux/Mac:  /Users/astro/code/my-project/src/pages/index.astro
+ * Windows:    C:/Users/astro/code/my-project/src/pages/index.astro
+ */
+export function viteifyURL(filePath: URL): string {
+  return slash(fileURLToPath(filePath));
```
Comment (drwpow) [inline packages/astro/src/core/util.ts:84]: Change to the `viteifyURL` function: before, we were prefixing everything with `/@fs`. I found removing this, and providing this format to have a better success rate in matching with Vite’s module graph.

If this causes any problems let me know. But I couldn’t find any instance of us requiring `/@fs/*` in Astro.

File: packages/astro/src/vite-plugin-build-html/index.ts [hunk 2]
```diff
@@ -140,7 +137,7 @@ export function rollupPluginAstroBuildHTML(options: PluginOptions): VitePlugin {
           for (let node of findAssets(document)) {
             if (isBuildableLink(node, srcRoot)) {
               const href = getAttribute(node, 'href')!;
-              const linkId = viteifyPath(href);
+              const linkId = `/@fs${href}`;
```
Comment (drwpow) [inline packages/astro/src/vite-plugin-build-html/index.ts:140]: I didn’t want to affect the build with the “viteify“ change, so I manually left this as it was. cc @matthewp what do you think?

File: packages/astro/src/vite-plugin-astro/index.ts [hunk 3]
```diff
@@ -121,16 +125,6 @@ ${err.url}`;
           err.stack = `    at ${id}`;
         }
 
-        // improve esbuild errors
-        if (err.errors && tsResult?.map) {
```
Comment (drwpow) [inline packages/astro/src/vite-plugin-astro/index.ts:125]: This code was trying to tap into Vite’s error overlay for errors. It ended up not being used.

File: packages/astro/src/core/ssr/index.ts [hunk 4]
```diff
@@ -269,9 +296,9 @@ async function getHmrScript() {
 export async function ssr(ssrOpts: SSROptions): Promise<string> {
   try {
     const [renderers, mod] = await preload(ssrOpts);
-    return render(renderers, mod, ssrOpts);
+    return await render(renderers, mod, ssrOpts); // note(drew): without "await", errors won’t get caught by errorHandler()
```
Comment (drwpow) [inline packages/astro/src/core/ssr/index.ts:299]: 💣 this blew everything up (not `await`-ing). Find by @FredKSchott!

File: packages/astro/src/vite-plugin-jsx/index.ts [hunk 5]
```diff
@@ -137,6 +137,7 @@ export default function jsx({ config, logging }: AstroPluginJSXOptions): Plugin
         const { code: jsxCode } = await esbuild.transform(code, {
           loader: getEsbuildLoader(path.extname(id)) as esbuild.Loader,
           jsx: 'preserve',
+          sourcefile: id,
```
Comment (drwpow) [inline packages/astro/src/vite-plugin-jsx/index.ts:140]: Big fix in React/Preact/Solid: when we pass the filename to esbuild, it generates a much more helpful stacktrace

File: packages/renderers/renderer-vue/server.js [hunk 6]
```diff
@@ -1,5 +1,5 @@
-import { renderToString } from '@vue/server-renderer';
 import { h, createSSRApp } from 'vue';
+import { renderToString } from 'vue/server-renderer';
```
Comment (drwpow) [inline packages/renderers/renderer-vue/server.js:2]: Minor change: I was seeing why the sourcemaps weren’t working in the Vue renderer at all, and it’s because we’re compiling to HTML, which doesn’t really get a good sourcemap. We’ll have to tackle this as a later ticket.

But in digging into it a little, Vue recommends this dependency cleanup so I just did it (this saves us from having to keep 3 package versions in sync)

File: packages/astro/src/core/ssr/index.ts [hunk 7]
```diff
@@ -88,25 +88,54 @@ async function resolveRenderers(viteServer: vite.ViteDevServer, astroConfig: Ast
...
-  const anyError = e as any;
-  if (anyError.errors) {
+  if (Array.isArray((e as any).errors)) {
     const { location, pluginName, text } = (e as BuildResult).errors[0];
     const err = e as SSRError;
     if (location) err.loc = { file: location.file, line: location.line, column: location.column };
-    const frame = codeFrame(await fs.promises.readFile(filePath, 'utf8'), err.loc);
-    err.frame = frame;
+    let src = err.pluginCode;
+    if (!src && err.id && fs.existsSync(err.id)) src = await fs.promises.readFile(err.id, 'utf8');
```
Comment (drwpow) [inline packages/astro/src/core/ssr/index.ts:113]: Improvement: check `pluginCode` for a better code frame in some cases

File: packages/astro/src/core/ssr/index.ts [hunk 8]
```diff
@@ -88,25 +88,54 @@ async function resolveRenderers(viteServer: vite.ViteDevServer, astroConfig: Ast
...
+    const err = e as SSRError;
+    const [message, maybeLoc] = err.stack.split('\n');
+    const fileMatch = maybeLoc.match(/at[^\(]+\(([^\)]+)/);
+    if (fileMatch && fileMatch[1]) {
+      let [srcLoc, lineStr, columnStr] = fileMatch[1].split(':');
+      let line = lineStr ? parseInt(lineStr, 10) : undefined;
+      let column = columnStr ? parseInt(columnStr, 10) : undefined;
+      if (fs.existsSync(srcLoc) && typeof line === 'number' && typeof column === 'number') {
+        err.loc = { file: srcLoc, line, column };
+        err.frame = codeFrame(await fs.promises.readFile(srcLoc, 'utf8'), err.loc);
```
Comment (drwpow) [inline packages/astro/src/core/ssr/index.ts:132]: Improvement: render code frame for more errors, when we can

File: packages/astro/src/core/ssr/index.ts [hunk 9]
```diff
@@ -228,7 +256,7 @@ export async function render(renderers: Renderer[], mod: ComponentInstance, ssrO
       // HACK: inject the direct contents of our `astro/runtime/client/hmr.js` to ensure
       // `import.meta.hot` is properly handled by Vite
       children: await getHmrScript(),
-      injectTo: 'head',
+      injectTo: 'head-prepend',
```
Comment (drwpow) [inline packages/astro/src/core/ssr/index.ts:259]: Unrelated change: Vite’s HMR gets injected at the very beginning of `<head>` so I figured we’d do the same.

File: packages/astro/src/vite-plugin-build-html/index.ts [hunk 10]
```diff
@@ -140,8 +137,7 @@ export function rollupPluginAstroBuildHTML(options: PluginOptions): VitePlugin {
           for (let node of findAssets(document)) {
             if (isBuildableLink(node, srcRoot)) {
               const href = getAttribute(node, 'href')!;
-              const linkId = viteifyPath(href);
-              assetImports.push(linkId);
+              assetImports.push(href);
```
Comment (drwpow) [inline packages/astro/src/vite-plugin-build-html/index.ts:140]: @matthewp since the change to the `viteifyURL()` function, I believe the `href`s are already correctly-formatted and Vite knows how to resolve them. Is this correct?

File: packages/astro/test/astro-styles-ssr.test.js [hunk 11]
```diff
@@ -2,27 +2,25 @@ import { expect } from 'chai';
 import cheerio from 'cheerio';
 import { loadFixture } from './test-utils.js';
 
-describe('Styles SSR', function () {
-  this.timeout(5000);
```
Comment (drwpow) [inline packages/astro/test/astro-styles-ssr.test.js:6]: This was left over from debugging and I’m surprised no one else hit it yet 🤦🏻

File: packages/astro/src/core/ssr/index.ts (hunk 4 above)
Comment (matthewp) [reply packages/astro/src/core/ssr/index.ts:299]: oof, good catch

**Comment by matthewp** [comment]
Looks great! Will want a changeset.

**Comment by matthewp** [comment]
Did some testing and found a case where an error wasn't reported due to a bug in the error handler, so went ahead and fixed it.

**Comment by natemoo-re** [comment]
@drwpow Looks great! Dev server test infra was just merged from https://github.com/snowpackjs/astro/pull/1879 if you want to add any tests to this PR.

**Comment by drwpow** [comment]
> Did some testing and found a case where an error wasn't reported due to a bug in the error handler, so went ahead and fixed it.

@matthewp Good catch! Thank you.

**Comment by jonathantneal** [comment]
Resolves https://github.com/snowpackjs/astro/issues/1890

**Comment by drwpow** [comment]
wip: adding tests & breaking some things out into another PR for easier review

File: packages/astro/src/vite-plugin-jsx/index.ts [hunk 12]
```diff
@@ -137,6 +137,8 @@ export default function jsx({ config, logging }: AstroPluginJSXOptions): Plugin
...
+          sourcemap: 'inline',
```
Comment (FredKSchott) [inline packages/astro/src/vite-plugin-jsx/index.ts:141]: have you confirmed that this fixed the issue / are JSX stack traces now more accurate with this fix?

Comment (drwpow) [reply packages/astro/src/vite-plugin-jsx/index.ts:141]: Yes! I was seeing the correct line/column on JSX errors now

**Review by FredKSchott — APPROVED** [review]
LGTM!
